### PR TITLE
Bedrock: share more test utils, cleanup derive package structure, move payload/attributes types to eth package

### PR DESCRIPTION
**Pull request by protolambda**
**No changes or additions in functionality in this PR**. These changes aim to clean up the structure/naming of the `derive` package in preparation of the upcoming batch derivation changes, and usage in Cannon.

Changes:
- Move `ExecutionPayload`, `PayloadAttributes`, and related API response types to the `eth` package.
  - This enables us to use them in the test utils, and pull the test utils into every package (except package `eth` itself), without circular dependencies
  - Cannon can implement the `Engine` interface that the driver code uses, to substitute the external engine with an in-memory EVM (with pre-image-oracle based state db) to run the fraud proof with.
- Split up `payload_attributes.go`: it didn't even contain the payload attributes type, and it was a mixed bag of things.
- The `GenerateDepositLog` function is now `MarshalDepositLogEvent`, since it nicely mirrors `UnmarshalDepositLogEvent` (slight rename from previous `UnmarshalLogEvent`) and we need it to be exposed for testing in other packages. This can't be part of the test util package, since the test-util package does not pull in anything from `derive` (since it's used for testing `derive` itself)
- Move random data generators from test files to test utils package
- Move `testID` (to easily create `eth.BlockID` with in tests) to test utils package
- Minor renames / typo fixes / related doc updates.

### CODE REVIEW

File: op-e2e/system_test.go [hunk 1]
```diff
@@ -10,9 +10,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/ethereum-optimism/optimism/op-node/eth"
```
Comment (trianglesphere) [inline op-e2e/system_test.go:13]: your IDE is still doing weird go import things. It's maybe worth figuring out an auto formatter for imports (gofmt only groups imports inside the block)

**Review by trianglesphere — APPROVED** [review]
Restructure looks good to me.
Only weird thing are some of the imports. I'm fine addressing them now w/out needing re-review or addressing them in a follow up PR.

File: op-node/rollup/derive/doc.go [hunk 2]
```diff
@@ -3,14 +3,13 @@
 // turned back into L1 data.
 //
 // The flow is data is as follows
-// receipts, batches -> l2.PayloadAttributes with `payload_attributes.go`
-// l2.PayloadAttributes -> l2.ExecutionPayload with `execution_payload.go`
-// L2 block -> Corresponding L1 block info with `l1_block_info.go`
+// receipts, batches -> eth.PayloadAttributes, by parsing the L1 data and deriving L2 inputs
+// l2.PayloadAttributes -> l2.ExecutionPayload, by running the EVM (using an Execution Engine)
+// L2 block -> Corresponding L1 block info, by parsing the first deposited transaction
```
Comment (norswap) [inline op-node/rollup/derive/doc.go:8]: Should we indicate the functions used to performe the transformations?

**Review by norswap — APPROVED** [review]
Looks good to me, definitely a step in the right direction!

File: op-node/rollup/derive/doc.go (hunk 2 above)
Comment (protolambda) [reply op-node/rollup/derive/doc.go:8]: We should long-term, but they're changing with remaining batch-derivation changes, so I'm going for something more generic that doesn't break.

File: op-node/testutils/block_id.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"strconv"
+	"strings"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type TestID string
+
```
[the remark this answers is not in the document]
Comment (protolambda) [reply op-node/testutils/block_id.go:12]: Will update. Mergify will auto-merge after this comment is resolved, so I'll leave it open before pushing the commit.

**Pull request merged**